### PR TITLE
distribute rubocop as a gem

### LINKED ISSUE #70: distribute rubocop styles as a gem
right now all our projects fetch our shared rubocop guide directly via HTTP. see [our rubocop README](https://github.com/tedconf/code-style-guides/blob/master/linters/rubocop/README.md#usage) for more details.

## pros to our current approach

new style rules are applied to all projects immediately. no update has to happen in each project to get the latest rules.

## cons for our current approach

rubocop changes a lot. "existing cops getting renamed" is especially problematic. 

for example, if i update a project to use the latest rubocop (0.72.0 as of 2019-07-03), i get the following error

```
   Error: The `Layout/FirstParameterIndentation` cop has been renamed to `Layout/IndentFirstArgument`.
(obsolete configuration found in .rubocop-https---raw-githubusercontent-com-tedconf-code-style-guides-master-linters-rubocop-rubocop-yml, please update it)
```

in order to make our shared guide work, i need to stick to a rubocop release which is prior to this rename. (or i could rename the cop in our style guide, but that breaks everybody who has an old version - with no warning/explanation.)

## proposal

  1. package our shared style guide as a gem like `ted-rubocop-styles`
  1. make that gem depend on a compatible rubocop version
  1. have projects depend on `ted-rubocop-styles` rather than on `rubocop` directly

this does mean all our projects need to update whenever we release a new version of our gem. i don't love that, but i don't see a cleaner way to solve the errors which come with cop renaming.

definitely open to other suggestions.

@tedconf/backenders thumbs up/down on this one?

### PR DESCRIPTION
**Pull request by alexdean**
fixes #70. this PR packages our rubocop styles as a gem. 

## feedback requested

@tedconf/backenders I'd love some input on how this approach seems to everybody.

  1. do my installation instructions (below) work for you? if not, what went wrong?
  2. if you wanted to make a change to the rubocop rules, do you know how you'd go about that? (i can/will improve the docs on that if needed.)
  3. once we're packaging our rubocop rules as a gem, our projects won't all automatically get the latest styles (as we currently do since we're all sourcing rules from master over HTTP). projects will have to explicitly update the `ted_rubocop_rules` gem to get updated rules. does that sound ok?

## TODO

  - [x] our existing process for sourcing the rule set via HTTP (from raw.githubusercontent.com) should continue to work, but that needs a bit of verification. (i'm trying a symlink to avoid copy/pasting the file. not 100% sure that'll play nice with github.com.)
  - [x] better docs on how to do a new release
  - [ ] get a few projects to install and use a prerelease of this gem. (notes below.)

## how to test this out

1. add this to a project's Gemfile

```ruby
gem 'ted_rubocop_rules'
```

2. `bundle update --conservative rubocop` to make sure you have the version of rubocop required our gem.

3. update your `.rubocop.yml`

```yml
AllCops:
  TargetRubyVersion: 2.5 # adjust for your project

inherit_gem:
  ted_rubocop_rules: rubocop.yml

inherit_from: .rubocop_todo.yml # if you have one
```

4. regenerate your todo list. `bundle exec rubocop --auto-gen-config`

### CODE REVIEW

**Comment by griley** [comment]
>  projects will have to explicitly update the ted_rubocop_rules gem to get updated rules. does that sound ok?

much preferred. thank you for that!

**Comment by DanielRussell** [comment]
Mentioning for completeness: since `code-style-guides` is a public repo, so we could have people install the gem straight from GH.  This is not great from an uptime-dependency (GH vs our own control) or repeatability (it's a lot easier to change a tag than republish a gem without updating the version) perspective.

But it would avoid the need for all builds to have access to our private gemserver.  In practice this is likely not a problem, but I wonder if there will be some unexpected edge cases.

**Comment by DanielRussell** [comment]
Should the instructions tell the user to remove `rubocop` from their `Gemfile`?  It'll work either way, but I think the intent is for the gem to pull rubocop as a dependency.

**Comment by DanielRussell** [comment]
The "how to test this out" instructions in this PR don't match the instructions in the style guide.  I'm not sure which one you want, so I've left them as-is.

I think it's smart to put this in a `group` in the `Gemfile`, tho.

**Comment by DanielRussell** [comment]
re: "make a change to the rubocop rules".  The instructions to edit the .yml and create a PR are clear.  I don't know if you have any preferences for required reviews or who should do a merge and release to ensure consistency.  There are no release instructions, I assume that's the standard steps, but a few words would help.

It might be worth a note about post-release steps (Announce it?  Don't bother?)

Based on the way this is laid out, it appears that if we ever want to change the styleguide (e.g. to across-the-board bump ABC from 25 to 60), only the latest gem will get it.  IOW: we're not planning to maintain a living style guide for multiple versions of rubocop.  I think this is a sensible idea, but don't know if it's worth specifically mentioning.

**Comment by DanielRussell** [comment]
Is it worth adding an upgrade guide similar to the install guide (e.g. upgrade the gem and regenerate your rubocop-todo.yml)?  Or wait to do that until we have the first round of upgrades, so we know what to write there?

File: gem/README.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+Individual projects may further extend this, but generally the fewer exceptions
+we define the better.
+
+## Usage
+
+in your `Gemfile`
+
+```ruby
+group :development, :test do
+  gem 'ted_rubocop_rules'
```
Comment (jhirbour) [inline gem/README.md:18]: should this reference installing the gem via the ted gem server instead?

Comment (alexdean) [reply gem/README.md:18]: no. see a69fe8ece2211f06c830ee8803d1561eb8652506.

**Comment by alexdean** [comment]
@DanielRussell I forgot this is a public repo. Since it is, we can publish the gem to rubygems.org rather than to rubygems.ted.com. I've made that change, and included some build instructions. Make sense?

To test latest: update your `Gemfile` to use just `gem 'ted_rubocop_rules'` (not `rubygems.ted.com`.) then `bundle update ted_rubocop_rules` and lmk how it goes.

**Pull request merged**